pandas/core/generic.py

```diff
@@ -3124,7 +3124,12 @@ def fillna(self, value=None, method=None, axis=None, inplace=False,
                 # fill in 2d chunks
                 result = dict([(col, s.fillna(method=method, value=value))
                                for col, s in self.iteritems()])
-                return self._constructor.from_dict(result).__finalize__(self)
+                if inplace:
+                    new_obj = self._constructor.from_dict(result).__finalize__(self)
+                    self._update_inplace(new_obj._data)
+                    return
+                else:
```

Comment:
return is necessary, because otherwise it will continue to run to the default code for "2d or less".
But I can refactor this part so that all changes to _data will be put into "new_data" and handled by downstream code.